Support S3 Server Side Encryption with custom key

tuxgasy opened the pull request:
Add a supports of Server Side Encryption known as [SSE-C](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html) for compatible S3 bucket provider.

The encryption is performed on the S3 server with the provided key. The key is not stored on the S3 server. To download encrypted file, you must therefore provide the same key.

The encryption key must be a base64 encoded string with a length of 32 bytes. You can generate one as following:

`openssl rand 32 | base64`

tuxgasy committed:
Add S3 Server side encryption with custom key (SSE-C)

tuxgasy commented:
MD5 hash of the encryption key is required. How to ignore the failing check ?

MichaIng commented on motioneye/uploadservices.py:
```diff
@@ -1221,10 +1226,20 @@ def upload_file(self, target_dir, filename, camera_name):
+            extra_args = {
+                "SSECustomerAlgorithm": "AES256",
+                "SSECustomerKey": self._sse_c_key,
+                "SSECustomerKeyMD5": sse_key_md5,
```
Is it possible to pass the key with a stronger hashing algorithm?

MichaIng replied:
Oh it actually seems this is not needed but calculated automatically if missing: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/multipartupload/complete.html
> Please note that this parameter is automatically populated if it is not provided. Including this parameter is not required

MichaIng commented on motioneye/uploadservices.py:
```diff
@@ -1221,10 +1225,23 @@ def upload_file(self, target_dir, filename, camera_name):
+        s3.upload_file(
+            filename,
+            self._bucket,
+            filename[(len(target_dir)+1) :],
```
The `+1` was added to remove the last slash as well? Did this always appear in the resulting key? But what happens if the `target_dir` is passed with trailing slash? I am currently not sure whether it is striped somewhere, otherwise we could either `target_dir.rstrip('/')` or alternatively `filename[len(target_dir) :].lstrip('/')` to achieve the same effect.

MichaIng replied:
Ohh, it seems to be actually solved better here: https://github.com/motioneye-project/motioneye/pull/2760/changes
Let me merge this first.

tuxgasy commented:
> @tuxgasy Can you please test it with the `SSECustomerKeyMD5` argument removed? As said, according to boto3 docs, it is not needed, but calculated automatically if missing. And see my above question regarding `+1` removed characters for the key name.

Without SSECustomerKeyMD5, I get the following error message :

```
boto3.exceptions.S3UploadFailedError: Failed to upload /var/lib/motioneye/Camera1/2026-02-04/18-07-57.jpg to me-salon/2026-02-04/18-07-57.jpg:
An error occurred (InvalidArgument) when calling the PutObject operation: The secret key was invalid for the specified algorithm.
```

MichaIng commented:
This is latest boto3? Weird, the docs say it is not needed: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/upload_part.html
Though this is `upload_part` and a bunch of other methods: https://boto3.amazonaws.com/v1/documentation/api/latest/search.html?q=SSECustomerKeyMD5
But the `upload_file` page does not explicitly mention it: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/upload_file.html

However, if it does not work reliably without, I'll add it back in.

Since I want to fix/merge #2760 first, can you confirm that using slashes in the filename creates subdirectories on the S3 server?

tuxgasy committed:
Fix bug when remove S3 SSE-C key

MichaIng committed:
minor code alignments

Signed-off-by: MichaIng <micha@dietpi.com>

MichaIng commented:
Yes, fixing this was the purpose of the other PR. Fixing the possible leading slashes was a side effect. The PR has been merged and this one rebased.

I restored the MD5 sum (and found the correct way to let `bandit` ignore the `md5()` call as weak algorithm 😅). Can you do a final test please?

MichaIng committed:
enh(main.html): extend S3 encryption info

Add a placeholder text which makes clear that the key is optional. Extend the tooltip text to make clear that this key is needed to access the files.

Signed-off-by: MichaIng <micha@dietpi.com>

MichaIng reviewed (approved):
Looks and works good for me now.
